### PR TITLE
Automated cherry pick of #105983: upgrade sigs.k8s.io/structured-merge-diff to 4.2.0

### PR DESCRIPTION
**Pull request by jiahuif**
Cherry pick of #105983 on release-1.22.

#105983: upgrade sigs.k8s.io/structured-merge-diff to 4.2.0

For details on the cherry pick process, see the [cherry pick requests](https://git.k8s.io/community/contributors/devel/sig-release/cherry-picks.md) page.

/kind feature

For details on the cherry pick process, see the [cherry pick requests](https://git.k8s.io/community/contributors/devel/sig-release/cherry-picks.md) page.

```release-note
kube-apiserver: when merging lists, Server Side Apply now prefers the order of the submitted request instead of the existing persisted object
```

### CODE REVIEW

File: test/integration/client/client_test.go [hunk 1]
```diff
@@ -1366,12 +1366,12 @@ func TestExtractModifyApply_ForceOwnership(t *testing.T) {
...
 					WithContainers(
-						corev1ac.Container().
-							WithName("nginx").
-							WithWorkingDir("/tmp/v2"),
 						corev1ac.Container().
 							WithName("sidecar").
 							WithImage("nginx:1.14.3"),
+						corev1ac.Container().
+							WithName("nginx").
+							WithWorkingDir("/tmp/v2"),
```
Comment (liggitt) [inline test/integration/client/client_test.go:1374]: previously, applying a non-overlapping item in a list appended, and now it prepends? that's surprising... I thought the change was only modifying preferring the request order when it specified items already present

Comment (liggitt) [reply test/integration/client/client_test.go:1374]: I missed that in the master PR... let's resolve this question in master before merging the cherry picks

**Comment by jiahuif** [comment]
Here is an explanation of the concat behavior change. Now that we prefer the ordering of RHS, everything that are not in RHS would have no defined order and thus may end up in any order. We would put them at the end. Before this change the same thing happened to items in RHS but not in LHS, new items would have no preferred position and be put at the end, which was what the original bug report. The ordering of disjointed items were not defined at the first place, and thus we just put it at the end for the ease of implementation.

File: test/integration/client/client_test.go (hunk 1 above)
Comment (jiahuif) [reply test/integration/client/client_test.go:1374]: This is a side effect of preferring the RHS order. The implementation would append everything without defined ordering to the result. Before the change, everything in LHS has defined order and thus each items in disjointed RHS would be appended. Now the situation reversed between LHS and RHS. Everything in disjointed LHS would be appended instead.

Comment (liggitt) [reply test/integration/client/client_test.go:1374]: I don't think the new behavior is correct either.

This is what happens now:
* apply `a,b,c` as fieldManager `x`
* apply `c,d,e,f` as fieldManager `y`
* final order is `c,d,e,f,a,b`

That discards the order expressed by fieldManager `x`, which is unexpected

Comment (liggitt) [reply test/integration/client/client_test.go:1374]: moving discussion to the original issue at https://github.com/kubernetes/kubernetes/issues/104641#issuecomment-976947921

**Comment by dims** [comment]
this PR has been on hold for about 2 weeks. Can you please see how to unblock this by reviewing comments etc? or please close this out (Trying to declutter our open list of PRs)

**Comment by liggitt** [comment]
/close

the original issue is still being worked on (xref https://github.com/kubernetes/kubernetes/issues/104641#issuecomment-978305017) and the final implementation can be picked in a new backport PR

**Pull request closed without merging**